### ISSUE TITLE
SWATEventIDs event matching method does not work very well

### ISSUE THREAD

**Issue by jurysek** [SST-1M-collaboration/sst1mpipe#61]
Looking at the last DL2 stereo data, taken on 20240806, I see that there is quite a lot of stereo events missing when we use "SWATEventIDs" event matching method. See the attached figure (DL2, after some gammaness cuts, event rates), which clearly illustrates the issue. Note that the issue might be there for quite a long time. When Adam introduced the SWATEventIDs, I only tested for some random set of files that the number of stereo events found is roughly the same as for the "WhiteRabbitClosest" method.

<img width="2265" alt="image" src="https://github.com/user-attachments/assets/ca1364b4-252f-4d5a-bc9b-dd19522a5410">

In the figure below, you can see the very same dataset, but in this case the stereo rate results from the "WhiteRabbitClosest" method. So here everything works fine and the issue must be then somewhere in the SWAT stereo event tagging.

<img width="2273" alt="image" src="https://github.com/user-attachments/assets/22586499-9ddd-451d-8688-38b2f2c262ef">

Quickly looking at the trigger rates below, it seems that SWAT breaks when the trigger threshold is changed, or if there is a spike in the rate.

<img width="1179" alt="image" src="https://github.com/user-attachments/assets/8569e5fc-7a6d-4a43-9904-fbac20baa13b">

@adam-muraczewski , could you please take a look?

**Comment by jurysek** [comment]
@adam-muraczewski OK, good! Yes, exactly, the data are from the night of 20240806. I do not have access to the list of failed files, because I re-wrote the stereo DL1 with those processed with the "WhiteRabbitClosest" option, sorry for that.. But the numbers on the x axis are the UNIX times, stored in the local_time column of the DL1 files.

**Comment by adam-muraczewski** [comment]
@jurysek I've plotted the same thing using raw R1-like data from SWAT's log for the night and the stereo trigger rate seems to be constant at roughly 70 triggers per second, so it's not the case that SWAT failed during that night:

![image](https://github.com/user-attachments/assets/c12510d5-23f6-408e-8259-027bc1541e53)

What I'm thinking now is maybe our window of 400 ns is too wide? If it is the case maybe SWAT matched triggers from 2 subsequent events as it uses the earliest events that meet the requirement and not the closest time-wise as your post-processing method does. Would such mismatch be detected by the quality cut that you've applied?

**Comment by adam-muraczewski** [comment]
Another test that I plan to do is to verify that the SWAT log matches with data from R1 fits, but for that I need calculus to not to download 1TB of fits files. Simple manual inspeciton of the end and beggining of the night I did yesterday showed no deviations, but I want to be sure.

**Comment by jurysek** [comment]
> What I'm thinking now is maybe our window of 400 ns is too wide? If it is the case maybe SWAT matched triggers from 2 subsequent events as it uses the earliest events that meet the requirement and not the closest time-wise as your post-processing method does. Would such mismatch be detected by the quality cut that you've applied?
Yes, if the event from a single telescope does not make it through cleaning, it is not in the stereo distribution above. This may indeed happen if one of the telescopes is triggered with NSB. But it is not clear to me how it would cause what we observe, as the trigger rate is mostly stable during the night except for some short NSB bursts. The event ids stop matching roughly at the time of the burst and recover later.

**Comment by tt-thomas** [comment]
400 ns is very small compared to 70 Hz, this is very likely not the issue here.
I think what happend is  some shift in the event IDs. This probably append when some buffer is full (we see the lost of coincidence match burst in trigger rate).
is the SWATEventID direcly sent by the SWAT, or is it something like the camerserver that increment the SWATEventID each time it receive a stereo flag?
if the data/IDs sent by the SWAT are correct. Is it possible that the camera server read them wrong?
I cant do this right now, but I would like to compare the SWATEventID from tel 1&2 in the period where the coincidences are lost.

**Comment by adam-muraczewski** [comment]
@tt-thomas We need to look at mono trigger rates which are several times higher and "bursty", so IMO this could still be an issue, but remains to be seen.

To answer your questions about SWAT operation:
- The assigned array event IDs are directly sent by SWAT and forwarded to zfitswriter so that they are written to output zfits files (we can see that because a Camera Server reset doesn't reset the ID, only SWAT restart does, also the IDs seem to match between SWAT log and the zfits files).
- I find an error in communication/reading very unlikely: we use TCP to ensure that all data goes through without error and in order: the OS itself makes sure of that; if there was some problem related to parsing I'd expect it to be persistent (not intermittent) and there would be a shift of at least 8 bytes which would make that kind of an error very noticable.

That said, once calculus is back I'll check if all timestamps and IDs match for the whole night.

**Comment by adam-muraczewski** [comment]
@jurysek @tt-thomas I've compared the contents of the raw zfits files and the SWAT log and interestingly they are not the same which explains the loss of stereo for the "SWATEventIDs" method.

For example, the following is an excerpt from the SWAT log for the night that was generated during the first trigger rate peak at ~21:13:
```
arrseqnum mono/stereo tel seqnum s qns trigger_type
299099807 1 21 830644 1722978812 1066430720 0
299099808 1 21 830645 1722978812 1066639424 0
299099809 1 21 830648 1722978812 1067676992 0
299099810 1 21 830236 1722978812 1068062995 0
299099811 1 21 830238 1722978812 1068327763 0
299099812 1 21 830239 1722978812 1068630547 0
299099813 1 21 830649 1722978812 1068995136 0
299099814 1 21 830650 1722978812 1069188224 0
299099815 1 21 830651 1722978812 1069791040 0
```

And here is the corresponding data from the zfits file:
```
299099807 1 21 830644 1722978812 1066430720 0
299099808 1 21 830645 1722978812 1066639424 0
299099809 1 21 830648 1722978812 1067676992 0
299099810 1 21 830649 1722978812 1068995136 0
299099811 1 21 830650 1722978812 1069188224 0
299099812 1 21 830651 1722978812 1069791040 0
```

Notice how the 4th column shows that SWAT received 3 triggers with camera-assigned IDs out of order (830236 after 830648) that aren't there in the zfits file, but their SWAT-assigned array event IDs are used for subsequent triggers which creates an offset between the SWATEventID value that is used by sst1mpipe and one that SWAT actually assigned. The same situation happens repeatedly during the affected time periods increasing the offset but then is fixed without any intervetion to neither SWAT nor the Camera Server, so it's very likely that the problem was fixed by reconfiguring (resetting?) the Camera during normal wobble change.

Could you please look at the data from the night and see whether it has other evidence of the TEL1 Camera being desynchronized such as mismatched 1/3 of the image or non-continuous waveforms (like what we had at the beggining of the year)?

Outside of the 3 problematic periods the shape of the plots from the first entry in the thread is exactly the same for both methods which means that SWAT worked as expected then. Also, the phenomenon only affected Telescope 1 while Telescope 2 shows no evidence of such desynchronization.

I'm not sure how it is possible for the Camera Server - SWAT system to mismatch or alter the IDs yet, but given what we know it seems that this is somehow related to the Camera Server reacting to a desync in the hardware. I'm coordinating with Jurek to understand the issue better.
